x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java

```diff
@@ -0,0 +1,117 @@
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.features.FeatureService;
+import org.elasticsearch.persistent.AllocatedPersistentTask;
+import org.elasticsearch.persistent.PersistentTaskState;
+import org.elasticsearch.persistent.PersistentTasksExecutor;
+import org.elasticsearch.xpack.core.security.action.UpdateIndexMigrationVersionAction;
+import org.elasticsearch.xpack.core.security.support.SecurityMigrationTaskParams;
+
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Executor;
+
+import static org.elasticsearch.xpack.security.support.SecurityIndexManager.loadIndexMappingVersion;
+
+public class SecurityMigrationExecutor extends PersistentTasksExecutor<SecurityMigrationTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(SecurityMigrationExecutor.class);
+    private final SecuritySystemIndices securitySystemIndices;
+    private final Client client;
+    private final FeatureService featureService;
+    private final ClusterService clusterService;
+    private final TreeMap<Integer, SecurityMigrations.SecurityMigration> securityMigrationByMigrationVersion;
+
+    public SecurityMigrationExecutor(
+        String taskName,
+        Executor executor,
+        SecuritySystemIndices securitySystemIndices,
+        Client client,
+        FeatureService featureService,
+        ClusterService clusterService,
+        TreeMap<Integer, SecurityMigrations.SecurityMigration> securityMigrationByMigrationVersion
+    ) {
+        super(taskName, executor);
+        this.securitySystemIndices = securitySystemIndices;
+        this.client = client;
+        this.featureService = featureService;
+        this.clusterService = clusterService;
+        this.securityMigrationByMigrationVersion = securityMigrationByMigrationVersion;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SecurityMigrationTaskParams params, PersistentTaskState state) {
+        applyOutstandingMigrations(params.getMigrationVersion(), ActionListener.wrap((res) -> task.markAsCompleted(), (exception) -> {
+            logger.warn("Security migration failed: " + exception);
+            task.markAsFailed(exception);
+        }));
+    }
+
+    private void applyOutstandingMigrations(int currentMigrationVersion, ActionListener<Void> listener) {
+        if (securityMigrationByMigrationVersion.lastKey() > currentMigrationVersion) {
+            Map.Entry<Integer, SecurityMigrations.SecurityMigration> migrationEntry = securityMigrationByMigrationVersion.higherEntry(
+                currentMigrationVersion
+            );
+
+            int currentMappingsVersion = loadIndexMappingVersion(
```

Comment:
I think this can be checked more easily before triggering the migration.

In general, I don't have a good feeling what should we check before the task is triggered vs in the task itself. I need to think more about it.

Reply:
It needs to be checked per migration since we apply all migrations recursively. We can still make sure that we have at least one eligible migration before starting, but when iterating through the migrations we need to see if we've reached one that is not supported before applying it. This can happen in a cluster where the mapping version required has not yet been applied or in a mixed cluster where all nodes are not yet on the required node feature.